### PR TITLE
Itea 28 extend travel platform using existing functionality and izi travel api

### PR DESCRIPTION
**Pull request by kate-ka**

### CODE REVIEW

File: travel_platform/travel_app/views.py [hunk 1]
```diff
@@ -1,12 +1,45 @@
...
+
+    if not detail:
+        detail = find_museum_detail(id)
+        cache.set(id, detail)
+    return render(request, 'travel_app/museum_detail.html',
+                  context={'detail': detail})
+
+
+def tour_detail(request, id):
+    detail = cache.get(id)
```
Comment (UkrainianEngineer) [inline travel_platform/travel_app/views.py]: Hm..it looks like we may use decorator here..
It will return value from cache if it exists, execute function with calculation of this value and adding into cache otherwise.
What do you think about this?

This decorator may be used for all of these functions, which needs caching.

File: travel_platform/travel_app/views.py [hunk 2]
```diff
@@ -1,12 +1,45 @@
...
+from api.izi_travel.izi_travel_data import *
 
 
 def index(request):
     return render(request, 'travel_app/main.html')
+
+
+def get_city_tourist_info(request):
+    city = request.GET['desired_location']
+    info = cache.get(city)
```
Comment (UkrainianEngineer) [inline travel_platform/travel_app/views.py]: How do you feel working with caching? Is it difficult to understand?
Or it's clear enough?

File: travel_platform/travel_app/views.py [hunk 3]
```diff
@@ -1,12 +1,45 @@
...
+        cache.set(city, {"museums": museums, "tours": tours})
+        context = {'museums': museums, 'tours': tours, 'city': city}
+    else:
+        museums = info.get("museums")
+        tours = info.get("tours")
+        context = {'museums': museums, 'tours': tours, 'city': city}
+    return render(request, 'travel_app/search.html', context=context)
+
+
+def museum_detail(request, id):
```
Comment (UkrainianEngineer) [inline travel_platform/travel_app/views.py]: I suggest to use `id_` for example, because `id` is builtin function.

File: travel_platform/travel_app/templates/travel_app/tour_detail.html [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+{{ detail }}
+{% block content %}
+
+  {% for attraction in detail %}
+  <h1>{{ attraction.tour_title }}</h1>
+    <h3>{{ attraction.title }}</h3>
+
+    {% for image in attraction.images %}
+      <div>
+        <img src="{{ image }}">
```
Comment (UkrainianEngineer) [inline travel_platform/travel_app/templates/travel_app/tour_detail.html]: It's a good manner to add `alt` parameter.
So, if we may extract any description from `image` please add it.

Thanks!

**Comment by UkrainianEngineer** [comment]
One general comment:
General implementation is really very nice!
But, please improve a little bit styles for links, because they have similar color to background.

Thank you for your effort!

File: travel_platform/travel_app/views.py (hunk 2 above)
Comment (kate-ka) [reply travel_platform/travel_app/views.py]: First it was not clear enough how django.core.cache is connected with memcached backend but then i found out it in the caching framework source code, now it seems clear.

File: travel_platform/travel_app/views.py [hunk 5]
```diff
@@ -10,36 +11,24 @@ def index(request):
     return render(request, 'travel_app/main.html')
 
 
+@cache_page(86400)
```
Comment (UkrainianEngineer) [inline travel_platform/travel_app/views.py]: Which data will be cached via this decorator?

File: travel_platform/travel_platform/settings.py [hunk 6]
```diff
@@ -81,7 +81,13 @@
     }
 }
 
-
+CACHES = {
+    'default': {
+        'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
+        'LOCATION': '127.0.0.1:11211',
```
Comment (UkrainianEngineer) [inline travel_platform/travel_platform/settings.py]: Did you try to use `0.0.0.0:11211` instead?
If it works, leave it as mentioned above.

File: travel_platform/travel_app/views.py [hunk 7]
```diff
@@ -1,12 +1,33 @@
...
 
-from .models import Query
+from api.izi_travel.izi_travel_data import *
 
 
 def index(request):
     return render(request, 'travel_app/main.html')
+
+
+@cache_page(86400)
```
Comment (UkrainianEngineer) [inline travel_platform/travel_app/views.py]: It would be great to have TTL as configuration parameter instead of hard-coding.

File: travel_platform/api/izi_travel/izi_travel_client.py [hunk 8]
```diff
@@ -37,7 +37,7 @@ def _check_response_status(self, response):
                 response.status_code, response.content))
 
     def _prepare_image_url(self, content_provider_uuid,
-                           image_uuid, image_size='800x600'):
+                           image_uuid, image_size='480x360'):
```
Comment (UkrainianEngineer) [inline travel_platform/api/izi_travel/izi_travel_client.py:40]: What is the reason of this change?

File: travel_platform/travel_platform/settings.py (hunk 6 above)
Comment (kate-ka) [reply travel_platform/travel_platform/settings.py]: I didn't try, it works.

File: travel_platform/api/izi_travel/izi_travel_client.py (hunk 8 above)
Comment (kate-ka) [reply travel_platform/api/izi_travel/izi_travel_client.py:40]: To see smaller image on frontend I guess

File: travel_platform/travel_app/views.py (hunk 5 above)
Comment (kate-ka) [reply travel_platform/travel_app/views.py]: The caching key looks like url with GET parameters and response is its value is it suitabale for us??

File: travel_platform/travel_app/views.py (hunk 7 above)
Comment (kate-ka) [reply travel_platform/travel_app/views.py]: @pivancy  where should I add this conf parameter?

File: travel_platform/api/izi_travel/izi_travel_client.py (hunk 8 above)
Comment (UkrainianEngineer) [reply travel_platform/api/izi_travel/izi_travel_client.py:40]: Is it any sense to add an ability to see the original size of picture?It's just a suggestion.

Comment (kate-ka) [reply travel_platform/api/izi_travel/izi_travel_client.py:40]: I think it's pointless but i can just remove default image size anytime and get original pics.

File: travel_platform/travel_app/views.py (hunk 7 above)
Comment (kate-ka) [reply travel_platform/travel_app/views.py]: Added ttl to settings

File: travel_platform/api/izi_travel/izi_travel_client.py (hunk 8 above)
Comment (UkrainianEngineer) [reply travel_platform/api/izi_travel/izi_travel_client.py:40]: ok, it's just a suggestion for possible improvement.

File: travel_platform/travel_app/views.py (hunk 5 above)
Comment (UkrainianEngineer) [reply travel_platform/travel_app/views.py]: > A more granular way to use the caching framework is by caching the output of individual views. django.views.decorators.cache defines a cache_page decorator that will automatically cache the view’s response for you. It’s easy to use:

```
from django.views.decorators.cache import cache_page

@cache_page(60 * 15)
def my_view(request):
    ...
```
> cache_page takes a single argument: the cache timeout, in seconds. In the above example, the result of the my_view() view will be cached for 15 minutes. 

It looks like in our case we'll cache HTML which our functions returns.

File: travel_platform/travel_app/views.py (hunk 7 above)
Comment (UkrainianEngineer) [reply travel_platform/travel_app/views.py]: Ok, great!

**Comment by UkrainianEngineer** [comment]
Well, this feature looks good to me. :+1: 

One thing I suggest to do: replace `127.0.0.1` with `0.0.0.0` for caching backend, because we may have an issues with this in case of deployed service into remote host.

**Comment by RomanPryima** [comment]
@kate-ka 👍

**Comment by Sasha1152** [comment]
:+1:  Very good!
